src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs

```diff
@@ -0,0 +1,670 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Asn1;
+using System.Security.Cryptography.Asn1.Pkcs12;
+using System.Security.Cryptography.Asn1.Pkcs7;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal abstract class UnixPkcs12Reader : IDisposable
+    {
+        private const string DecryptedSentinel = nameof(UnixPkcs12Reader);
+
+        private PfxAsn _pfxAsn;
+        private ContentInfoAsn[] _safeContentsValues;
+        private CertAndKey[] _certs;
+        private int _certCount;
+
+        protected abstract ICertificatePalCore ReadX509Der(ReadOnlyMemory<byte> data);
+        protected abstract AsymmetricAlgorithm LoadKey(ReadOnlyMemory<byte> safeBagBagValue);
+
+        protected void ParsePkcs12(byte[] data)
+        {
+            // RFC7292 specifies BER instead of DER
+            AsnReader reader = new AsnReader(data, AsnEncodingRules.BER);
+            ReadOnlyMemory<byte> encodedData = reader.PeekEncodedValue();
+
+            // Windows compatibility: Ignore trailing data.
+            if (encodedData.Length != data.Length)
+            {
+                reader = new AsnReader(encodedData, AsnEncodingRules.BER);
+            }
+
+            PfxAsn.Decode(reader, out _pfxAsn);
```

Comment:
should you be storing out in the local? otherwise if the below check fails you still mutate the state despite the exception